extensions/rate_limiter.go

```diff
@@ -0,0 +1,95 @@
+package extensions
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+type rateLimiter struct {
+	redis    *redis.Client
+	rpmLimit int
+	l        *logrus.Entry
+}
+
+func NewRateLimiter(config *viper.Viper, logger *logrus.Logger) rateLimiter {
+	host := config.GetString("rateLimiter.redis.host")
+	port := config.GetInt("rateLimiter.redis.port")
+	pwd := config.GetString("rateLimiter.redis.password")
+	limit := config.GetInt("rateLimiter.limit.rpm")
+	isTest := config.GetBool("rateLimiter.test")
+
+	addr := fmt.Sprintf("%s:%d", host, port)
+	opts := &redis.Options{
+		Addr:     addr,
+		Password: pwd,
+	}
+	// Setting TLSConfig only for production due to not being able to enable TLS in the integration test container.
+	if !isTest {
+		opts.TLSConfig = &tls.Config{}
+	}
+
+	rdb := redis.NewClient(opts)
+
+	return rateLimiter{
+		redis:    rdb,
+		rpmLimit: limit,
+		l: logger.WithFields(logrus.Fields{
+			"extension": "RateLimiter",
+			"rpmLimit":  limit,
+		}),
+	}
+}
+
+// Allow checks Redis for the current rate a given device has in the current minute
+// If the rate is lower than the limit, the message is allowed. Otherwise, it is not allowed.
+// Reference: https://redis.io/glossary/rate-limiting/
+func (r rateLimiter) Allow(ctx context.Context, device string) bool {
+	deviceKey := fmt.Sprintf("%s:%d", device, time.Now().Minute())
+	l := r.l.WithFields(logrus.Fields{
+		"deviceKey": deviceKey,
+		"method":    "Allow",
+	})
+
+	val, err := r.redis.Get(ctx, deviceKey).Result()
+	if err != nil && !errors.Is(err, redis.Nil) {
+		// Something went wrong, return true to avoid blocking notifications.
+		l.WithError(err).Error("could not get current rate in redis")
+		return true
+	}
+	if errors.Is(err, redis.Nil) {
+		// First time
+		val = "0"
+	}
+
+	current, err := strconv.Atoi(val)
+	if err != nil {
+		// Something went wrong, return true to avoid blocking notifications.
+		l.WithError(err).Error("current rate is invalid")
+		return true
+	}
+
+	if current >= r.rpmLimit {
+		return false
+	}
+
+	_, err = r.redis.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+		pipe.Incr(ctx, deviceKey)
+		pipe.Expire(ctx, deviceKey, 1*time.Minute)
+		return nil
+	})
+	if err != nil {
+		// Allow the operation even if the transaction fails, to avoid blocking notifications.
```

Comment:
ditto here, metrics

Reply:
added it

Reply:
statsReporters is not being used to report here